### ISSUE TITLE
useLazyQuery doesn't change refetch function when the apollo client of the ApolloProvider changes

### ISSUE THREAD

**Issue by SimonMueller** [apollographql/apollo-client#11096]
### Issue Description

When changing the apollo client instance of the `ApolloProvider` context the refetch function provided by the `useLazyQuery` hook does not change and therefore an outdated version of the apollo client is used to make further requests.

It works as expected and the refetch function changes when using `useQuery` in the same way.

In our case the problem popped up when refreshing the auth tokens which we send in an http link for each request. After refreshing and using useLazyQuery we still see the old tokens as the refetch function is using the old apollo client and therefore the old auth link.

Versions:
- @apollo/client: 3.7.17
- graphql: 15.5.0

### Link to Reproduction

https://github.com/SimonMueller/apollo-client-context-change-lazy-query-bug-reproduction

### Reproduction Steps

1. Run `npm ci` to install
2. Run `npm start` to start the app
3. Open the app in the browser
4. Open the console tab of the dev tools in the browser
5. Click the "Make request" button and observe the client id logged to the console
6. Click the "Change client"
7. Click the "Make request" button again and observe the client id which has not changed but should have

**Comment by phryneas** [comment]
Good point, I can see that [in our code](https://github.com/apollographql/apollo-client/blob/30699ccc76cb07f5f4bc4959ade5b02fbd9091d1/src/react/hooks/useQuery.ts#L472-L476) - we should take a look at that.

Independently of that, I would *really* recommend against creating a new ApolloClient instance just because you changed an authentication token.  
You should keep the same ApolloClient instance over the lifetime of your application.  
Depending on how you created the client in the first place, you can usually share a variable in scope with the `setContext` Link and just modify that - so the link always uses the latest token available without having to recreate everything.

**Comment by SimonMueller** [comment]
wow that was a quick response! 😄 

and thanks for your note on the instance creation. makes total sense not only to minimize rerenders as the ApolloProvider is usually on the top of the hierarchy. Thats exactly how we fixed our problem. By using a reference to the latest token and not creating a new apollo client instance everytime.

I thought it would still be good to report it. 👍

**Comment by phryneas** [comment]
Absolutely, thanks a ton for the bug report!

**Comment by dyst5422** [comment]
Can confirm, using a reference is the way to go here - wildly simple and effective

**Comment by jerelmiller** [comment]
Hey @SimonMueller 👋 

Good news! Looks like this is fixed in the 4.0 alpha versions. We rewrote `useLazyQuery` to be more intuitive and in this process the function identities are correct. I tested your reproduction using the latest alpha build ([`4.0.0-alpha.19`](https://github.com/apollographql/apollo-client/releases/tag/%40apollo%2Fclient%404.0.0-alpha.19) at the time of this writing) and the app behaves as you expect where calling `execute` or `refetch` will use the updated client. This appears to be fixed as early as the first alpha [`4.0.0-alpha.0`](https://github.com/apollographql/apollo-client/releases/tag/v4.0.0-alpha.0). Please try the alpha version out and let us know if you have any feedback!

Since the fix is merged and released in the alpha version, I'm going to go ahead and close this issue. Let us know if you have any issues if you try out the new version. Thanks!

**Comment by SimonMueller** [comment]
@jerelmiller I just tried it with the versions you provided and everything works as expected now!
Thanks for letting me know and for fixing this 💯

**Comment by jerelmiller** [comment]
Wonderful! Great to hear its working on your end! Glad to be able to get this fixed finally!
